packages/apollo-react-relay-duct-tape/src/storeObservation/compiledHooks/useCompiledFragment.ts

```diff
@@ -28,73 +29,102 @@ export function useCompiledFragment(
     "useFragment(): Expected a `watchQueryDocument` to have been " +
       "extracted. Did you forget to invoke the compiler?",
   );
+  const from = fragmentReference.id;
+  invariant(
+    typeof from === "string",
+    "useFragment(): Expected the fragment reference to have a string id. " +
+      "Did you forget to invoke the compiler?",
+  );
+  invariant(
+    metadata,
+    "useFragment(): Expected metadata to have been extracted from " +
+      "the fragment. Did you forget to invoke the compiler?",
+  );
 
-  const client = useOverridenOrDefaultApolloClient();
-  const forceUpdate = useForceUpdate();
+  const defaultVariables = getDefaultVariables(watchQueryDocument);
+  const variables = {
+    ...defaultVariables,
+    ...fragmentReference.__fragments,
+  };
 
-  const observableQuery = useMemo(
-    () =>
-      client.watchQuery({
-        fetchPolicy: "cache-only",
-        query: watchQueryDocument,
-        returnPartialData: false,
-        variables: {
-          ...fragmentReference.__fragments,
-          id: fragmentReference.id,
-          __fragments: fragmentReference.__fragments,
-        },
-      }),
-    [client, fragmentReference.id, fragmentReference.__fragments],
-  );
+  const doc: UseFragmentOptions<unknown, unknown> = {
+    fragment: getFragmentDocumentFromWatchQueryDocument(
+      watchQueryDocument,
+      metadata,
+    ),
+    fragmentName: metadata?.mainFragment?.name,
+    from,
+    variables,
+  };
 
-  useEffect(() => {
-    let skipFirst = true;
-    const subscription = observableQuery.subscribe(
-      () => {
-        // Unclear why, but this yields twice with the same results, so skip one.
-        if (skipFirst) {
-          skipFirst = false;
-        } else {
-          forceUpdate();
-        }
-      },
-      (error) => {
-        console.log(error);
-      },
-    );
-    return () => subscription.unsubscribe();
-  }, [observableQuery]);
+  const result = useFragment(doc);
+  const data = result.complete ? result.data : {};
+  return data as object;
+}
 
-  const result = observableQuery.getCurrentResult();
-  if (result.partial) {
-    invariant(
-      false,
-      "useFragment(): Missing data expected to be seeded by the execution query document: %s. Please check your type policies and possibleTypes configuration. If only subset of properties is missing you might need to configure merge functions for non-normalized types.",
-      JSON.stringify(
-        // we need the cast because queryInfo and lastDiff are private but very useful for debugging
-        (
-          observableQuery as unknown as {
-            queryInfo?: { lastDiff?: { diff?: Cache.DiffResult<unknown> } };
-          }
-        ).queryInfo?.lastDiff?.diff?.missing?.map((e) => e.path),
-      ),
-    );
-  }
```

Comment:
We must keep this invariant, but trigger it if `useFragment` returns `!result.complete`

Reply:
thanks, fixed